plugins/outputs/bigquery/bigquery.go

```diff
@@ -0,0 +1,233 @@
+package bigquery
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/bigquery"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/outputs"
+)
+
+const (
+	defaultOffSetKey   = "offset-key.json"
+	timeStampFieldName = "timestamp"
+)
+
+var defaultTimeout = 5 * time.Second
+
+const sampleConfig = `	
+  ## Credentials File
+  credentials_file = "/path/to/service/account/key.json"
+  ## GCP Project
+  project = "my-gcp-project"
+
+  ## The namespace for the metric descriptor
+  dataset = "telegraf"
+
+  ## Timeout for BigQuery operations.  Minimum timeout is "1s".
+  # timeout = "5s"
+`
+
+type BigQuery struct {
+	CredentialsFile string `toml:"credentials_file"`
+	Project         string `toml:"project"`
+	Dataset         string `toml:"dataset"`
+
+	Timeout internal.Duration `toml:"timeout"`
+
+	Log telegraf.Logger
+
+	client *bigquery.Client
+}
+
+// SampleConfig returns the formatted sample configuration for the plugin.
+func (s *BigQuery) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description returns the human-readable function definition of the plugin.
+func (s *BigQuery) Description() string {
+	return "Configuration for Google Cloud BigQuery to send entries"
+}
+
+func (b *BigQuery) Connect() error {
+	b.setUpTimeout()
+
+	if b.Project == "" {
+		return fmt.Errorf("Project is a required field for BigQuery output")
+	}
+
+	if b.Dataset == "" {
+		return fmt.Errorf("Dataset is a required field for BigQuery output")
+	}
+
+	if b.client == nil {
+		return b.setUpDefaultClient()
+	}
+
+	return nil
+}
+
+func (b *BigQuery) setUpTimeout() {
+	if b.Timeout.Duration < 1*time.Second {
+		b.Timeout.Duration = defaultTimeout
+	}
+}
```

Comment:
Could you please also move this to `Connect()` to have a nice consistent structure. It will add to the beauty of this code.